nilearn/externals/skimage/random_walker_segmentation.py

```diff
@@ -0,0 +1,517 @@
+    return edges
+
+
+def _compute_weights_3d(data, spacing, beta=130, eps=1.e-6,
+                        multichannel=False):
+    # Weight calculation is main difference in multispectral version
+    # Original gradient**2 replaced with sum of gradients ** 2
+    gradients = 0
+    for channel in range(0, data.shape[-1]):
+        gradients += _compute_gradients_3d(data[..., channel],
+                                           spacing) ** 2
+    # All channels considered together in this standard deviation
+    beta /= 10 * data.std()
+    if multichannel:
+        # New final term in beta to give == results in trivial case where
+        # multiple identical spectra are passed.
+        beta /= np.sqrt(data.shape[-1])
+    gradients *= beta
+    weights = np.exp(- gradients)
+    weights += eps
+    return weights
+
+
+def _compute_gradients_3d(data, spacing):
+    gr_deep = np.abs(data[:, :, :-1] - data[:, :, 1:]).ravel() / spacing[2]
+    gr_right = np.abs(data[:, :-1] - data[:, 1:]).ravel() / spacing[1]
+    gr_down = np.abs(data[:-1] - data[1:]).ravel() / spacing[0]
+    return np.r_[gr_deep, gr_right, gr_down]
+
+
+def _make_laplacian_sparse(edges, weights):
+    """
+    Sparse implementation
+    """
+    pixel_nb = edges.max() + 1
+    diag = np.arange(pixel_nb)
+    i_indices = np.hstack((edges[0], edges[1]))
+    j_indices = np.hstack((edges[1], edges[0]))
+    data = np.hstack((-weights, -weights))
+    lap = sparse.coo_matrix((data, (i_indices, j_indices)),
+                            shape=(pixel_nb, pixel_nb))
+    connect = - np.ravel(lap.sum(axis=1))
+    lap = sparse.coo_matrix(
+        (np.hstack((data, connect)), (np.hstack((i_indices, diag)),
+                                      np.hstack((j_indices, diag)))),
+        shape=(pixel_nb, pixel_nb))
+    return lap.tocsr()
+
+
+def _clean_labels_ar(X, labels, copy=False):
+    X = X.astype(labels.dtype)
+    if copy:
+        labels = np.copy(labels)
+    labels = np.ravel(labels)
+    labels[labels == 0] = X
+    return labels
+
+
+def _buildAB(lap_sparse, labels):
+    """
+    Build the matrix A and rhs B of the linear system to solve.
+    A and B are two block of the laplacian of the image graph.
+    """
+    labels = labels[labels >= 0]
+    indices = np.arange(labels.size)
+    unlabeled_indices = indices[labels == 0]
+    seeds_indices = indices[labels > 0]
+    # The following two lines take most of the time in this function
+    B = lap_sparse[unlabeled_indices][:, seeds_indices]
+    lap_sparse = lap_sparse[unlabeled_indices][:, unlabeled_indices]
+    nlabels = labels.max()
+    rhs = []
+    for lab in range(1, nlabels + 1):
+        mask = (labels[seeds_indices] == lab)
+        fs = sparse.csr_matrix(mask)
+        fs = fs.transpose()
+        rhs.append(B * fs)
+    return lap_sparse, rhs
+
+
+def _mask_edges_weights(edges, weights, mask):
+    """
+    Remove edges of the graph connected to masked nodes, as well as
```

Comment:
I suggest we don't fix scikit image docstring. The goal is to be able to update this code with newer versions easily.